File: python/day-03/src/day_03/part_two.py
from dataclasses import dataclass
from typing import List, Tuple


@dataclass
class Point:
    x: int
    y: int
    val: str

    def vector_in_aoe(self, vec: "Vector") -> bool:
        if (self.y - 1) > vec.y or (self.y + 1) < vec.y:
            return False

        if (self.x - 1) > vec.x_max or (self.x + 1) < vec.x_min:
            return False

        return True


@dataclass
class Vector:
    x_min: int
    x_max: int
    y: int
    _val: str

    @classmethod
    def from_point(cls, point: Point):
        return cls(x_min=point.x, x_max=point.x, y=point.y, _val=point.val)

    def add_point(self, point: Point):
        assert point.y == self.y

        if point.x == self.x_min - 1:
            self.x_min = point.x
            self._val = point.val + self._val
        elif point.x == self.x_max + 1:
            self.x_max = point.x
            self._val += point.val

    @property
    def val(self):
        return int(self._val)


def read_input(fp: str) -> List[List[str]]:
    with open(fp, "r") as file:
        input = [line.strip() for line in file.readlines()]

    return input


def part_two(input: List[str]) -> int:
    gears, numbers = find_gears_and_numbers(input)

    gear_ratio_sum = 0
    for gear in gears:
        adjacent_parts = []
        for number in numbers:
            if gear.vector_in_aoe(number):
                adjacent_parts.append(number)

        if len(adjacent_parts) == 2:
            gear_ratio_sum += adjacent_parts[0].val * adjacent_parts[1].val

    return gear_ratio_sum


def find_gears_and_numbers(
    input: List[List[str]],
) -> Tuple[List[Point], List[Vector]]:
    gear_coordinates = []
    number_vectors = []

    for idx_y, line in enumerate(input):
        current_vector = None
        for idx_x, col in enumerate(line):
            if not col.isnumeric():
                if current_vector:
                    number_vectors.append(current_vector)
                    current_vector = None

                if col == "*":
                    point = Point(idx_x, idx_y, col)
                    gear_coordinates.append(point)

            if col.isnumeric():
                point = Point(idx_x, idx_y, col)
                if not current_vector:
                    current_vector = Vector.from_point(point)
                else:
                    current_vector.add_point(point)

        if current_vector:
            number_vectors.append(current_vector)
            current_vector = None

    return (gear_coordinates, number_vectors)


def main():
    input = read_input("input.txt")
    output = part_two(input)
    print(output)


if __name__ == "__main__":
    main()

File: rust/day-03/src/bin/part_two.rs
use std::cmp::min;

struct Point {
    x: u32,
    y: u32,
    val: String,
}

impl Point {
    fn vector_in_aoe(&self, vector: &Vector) -> bool {
        if self.y == 0 && (self.y + 1) < vector.y {
            return false;
        }

        if (self.y - 1) > vector.y || (self.y + 1) < vector.y {
            return false;
        }

        if self.x == 0 && (self.x + 1) < vector.x_min {
            return false;
        }

        if (self.x - 1) > vector.x_max || (self.x + 1) < vector.x_min {
            return false;
        }

        true
    }
}

#[derive(Debug, Clone, Copy)]
struct Vector {
    x_min: u32,
    x_max: u32,
    y: u32,
    val: u32,
}

impl Vector {
    fn from_point(point: Point) -> Self {
        Self {
            x_min: point.x,
            x_max: point.x,
            y: point.y,
            val: point.val.parse::<u32>().unwrap(),
        }
    }

    fn add_point(&mut self, point: Point) {
        assert_eq!(self.y, point.y);

        if min(0, self.x_min.wrapping_sub(1)) == point.x {
            self.x_min = point.x;
            self.val = format!("{0}{1}", point.val, self.val)
                .parse::<u32>()
                .unwrap();
        } else if self.x_max + 1 == point.x {
            self.x_max = point.x;
            self.val = format!("{0}{1}", self.val, point.val)
                .parse::<u32>()
                .unwrap();
        }
    }

    fn valid_point(&self, point: &Point) -> bool {
        if self.y != point.y {
            return false;
        }

        if min(0, self.x_min.wrapping_sub(1)) != point.x {
            if self.x_max + 1 != point.x {
                return false;
            }
        }

        true
    }
}

fn main() {
    let input = include_str!("./input.txt");
    let output = part_two(&input);
    dbg!(output);
}

fn part_two(input: &str) -> u32 {
    let gears = find_gears(input);
    let numbers = find_numbers(input);

    let mut gear_ratio_sum = 0;
    for gear in &gears {
        let mut adjacent_parts: Vec<Vector> = vec![];
        for number in &numbers {
            if gear.vector_in_aoe(number) {
                adjacent_parts.push(number.clone());
            }
        }

        if adjacent_parts.len() == 2 {
            let product = adjacent_parts[0].val * adjacent_parts[1].val;
            gear_ratio_sum += product;
        }

        adjacent_parts.clear();
    }

    gear_ratio_sum
}

fn find_gears(input: &str) -> Vec<Point> {
    let mut symbols = vec![];
    for (idx_y, line) in input.lines().enumerate() {
        for (idx_x, char) in line.chars().enumerate() {
            if char == '*' {
                let symbol = Point {
                    x: idx_x as u32,
                    y: idx_y as u32,
                    val: String::from(char),
                };
                symbols.push(symbol);
            }
        }
    }

    symbols
}

fn find_numbers(input: &str) -> Vec<Vector> {
    let mut numbers: Vec<Vector> = vec![];

    for (idx_y, line) in input.lines().enumerate() {
        for (idx_x, char) in line
            .chars()
            .enumerate()
            .filter(|(_, char)| char.is_ascii_digit())
        {
            let point = Point {
                x: idx_x as u32,
                y: idx_y as u32,
                val: String::from(char),
            };

            match numbers.last_mut() {
                Some(vec) => {
                    if vec.valid_point(&point) {
                        vec.add_point(point)
                    } else {
                        numbers.push(Vector::from_point(point))
                    }
                }
                None => numbers.push(Vector::from_point(point)),
            }
        }
    }

    numbers
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_part_two() {
        let result = part_two(
            "467..114..
...*......
..35..633.
......#...
617*......
.....+.58.
..592.....
......755.
...$.*....
.664.598..",
        );
        assert_eq!(result, 467835)
    }
}
